### ISSUE TITLE
Add RBS type signature

### ISSUE THREAD

**Issue by sferik** [kgiszczak/shale#38]
First, thank you for creating this project! Shale is a very nicely designed library and I've enjoyed working with it.

My only feature request is for you to add an [RBS](https://github.com/ruby/rbs) type signature file (`sig/shale.rbs`) to this project. I type check my code with [Steep](https://github.com/soutaro/steep) and now that many of my classes inherit from `Steep::Mapper`, my type checker is throwing dozens of errors.

If you'd like to see an example, here is an [`.rbs` file](https://github.com/sferik/x-ruby/blob/main/sig/x.rbs) and [`Steepfile`](https://github.com/sferik/x-ruby/blob/main/Steepfile) from one of my projects.

You may also find `typeprof` useful to generate RBS scaffolding.

**Comment by kgiszczak** [comment]
Would you be willing to take a stab at it? I'd gladly accept PR with this change. If not I'll try to add it myself, although it may take some time.

**Comment by sferik** [comment]
I forked this repo and created an [`rbs` branch](https://github.com/sferik/shale/tree/rbs) with [two commits](https://github.com/kgiszczak/shale/compare/master...sferik:shale:rbs):
1. [Add Steep to typecheck](https://github.com/sferik/shale/commit/f7f9ba92fd76f20a6a0a5fbd4109bb1eba680f5d)
2. [Automatically generate RBS file using TypeProf](https://github.com/sferik/shale/commit/9e4f5a1ce8884ca64819f72208fa9e5b3fac10e7)

When I run `bundle exec rake steep`, it detects 245 problems in the auto-generated type signatures. I'm not sure I have the expertise in this library necessary to fix them, but I figured I would update this issue with the scaffolding I generated in case that inspires someone else to attempt to resolve these problems.

I believe the majority of these issues can be resolved by adding `nil` checks or specifying that the value will not be `nil` in the type signature.
